lightautoml/transformers/utils.py

```diff
@@ -0,0 +1,269 @@
+"""utils for transformers."""
+
+import numpy as np
+
+from scipy.stats import mode
+
+
+def get_mode(x):
+    """Helper function to calculate mode."""
+    return mode(x)[0][0]
+
+
+class GroupByProcessor:
+    """Helper class to calculate group_by features."""
+
+    def __init__(self, keys):
+        super().__init__()
+
+        assert keys is not None
+
+        self.index, self.keys_as_int = np.unique(keys, return_inverse=True)
+        self.n_keys = max(self.keys_as_int) + 1
+        self.set_indices()
+
+    def set_indices(self):
+        """Sets indices for keys."""
+        self.indices = [[] for i in range(self.n_keys)]
+        for i, k in enumerate(self.keys_as_int):
+            self.indices[k].append(i)
+        self.indices = [np.array(elt) for elt in self.indices]
+
+    def apply(self, functions, vectors):
+        """Applies functions to vectors."""
+        assert functions is not None
+        assert vectors is not None
+
+        if isinstance(functions, list):
+            return [[fun(vec[idx].tolist()) for fun, vec in zip(functions, vectors)] for idx in (self.indices)]
+        else:
+            return [functions(vectors[idx].tolist()) for idx in (self.indices)]
+
+
+class GroupByFactory:
+    """Factory to create group_by classes."""
+
+    @staticmethod
+    def get_GroupBy(kind):
+        """Uses string identifiers to locate appropriate implementation.
+
+        Args:
+            kind: groupby feature transformation class
+
+        Example:
+            GroupByFactory.get_GroupBy('delta_mean')
+
+        Returns:
+            Object of GroupByBase impementing selected feature.
+
+        Raises:
+            ValueError: if identifier is not found.
+        """
+        assert kind is not None
+
+        available_classes = [
+            GroupByNumDeltaMean,
+            GroupByNumDeltaMedian,
+            GroupByNumMin,
+            GroupByNumMax,
+            GroupByNumStd,
+            GroupByCatMode,
+            GroupByCatIsMode,
+        ]
+
+        for class_name in available_classes:
+            if kind == class_name.class_kind:
+                return class_name(
+                    class_name.class_kind,
```

Comment:
it looks strange -- we don't need to pass fields from the class to its own __init__